Add keyboard shortcut: "V" to toggle between TextViews

khemarato opened the pull request:
This PR (built on top of https://github.com/suttacentral/suttacentral/pull/3236 so please ignore the extra string changes) takes a first stab at one of the keyboard shortcuts in https://github.com/suttacentral/suttacentral/issues/2308

khemarato committed:
bump client element localization files

khemarato committed:
Press V to toggle TextView

khemarato commented on client/elements/addons/sc-top-sheet-views.js:
```diff
@@ -183,6 +183,14 @@ export class SCTopSheetViews extends LitLocalized(LitElement) {
         color: var(--sc-on-tertiary-primary-text-color);
       }
+      
+      code.key {
```
This is what this looks like in both light and dark themes:

![Screenshot from 2024-08-04 14-29-08](https://github.com/user-attachments/assets/740a87cb-73fe-4d24-a407-5cd7ad942199)

![Screenshot from 2024-08-04 14-29-18](https://github.com/user-attachments/assets/af3f7659-545d-4b56-88ae-4bd852462bad)

Happy to change this styling to conform to whatever style guidelines you have.

khemarato commented on client/elements/addons/sc-top-sheet-views.js:
```diff
@@ -330,6 +338,31 @@ export class SCTopSheetViews extends LitLocalized(LitElement) {
+    window.setTimeout(() => {
+      this.show();
+      this.hide();
+    }, 2000);
```
This timeout is absolutely necessary to make it work, but I can't for the life of me figure out why.  Without this `show/hide`, the toast pops up showing that the key press was registered, but the layout doesn't actually change.  Any ideas, @ihongda ?

khemarato replied:
Thanks for playing with it @thesunshade ! Your bug is certainly the same one this is attempting to work around. It will definitely need a real fix before this can merge. That's waiting on @ihongda's knowledge: what gets hooked up only after the top panel shows on a given sutta?

ihongda replied:
> This timeout is absolutely necessary to make it work, but I can't for the life of me figure out why. Without this `show/hide`, the toast pops up showing that the key press was registered, but the layout doesn't actually change. Any ideas, @ihongda ?

Bhante 🙏🏼 

This part of the code can be placed in the `firstUpdated` event of lit.
>Called after the component's DOM has been updated the first time, immediately before [updated()](https://lit.dev/docs/components/lifecycle/#updated) is called.

In the firstUpdated event, we should be able to remove the `setTimeout`.

khemarato replied:
What are you talking about @ihongda? You mean put the addEventListener to the firstUpdated callback instead of the connectedCallback?  Cause that doesn't work.  I'm telling you that there's something that's only getting hooked up after the top sheet actually _opens_ (firstUpdated is called before that).  There's some function I need to call here (or some other logic somewhere else that I need to change?) to make sure that `actions.chooseSegmentedSuttaTextView` actually does what it's supposed to.

khemarato replied:
@ihongda - Okay. I figured it out. `sc-text-bilara.shouldRestoreUserSettings` is `false` by default and is suppressing the user-selected setting until the user opens the control panel, I assume to allow the url params to take effect without changing the user's settings.  In this case, the correct thing to do (?) is to set shouldRestoreUserSettings to `true` after the user hits one of the keyboard shortcuts.  Will update the PR with that logic.

khemarato commented on client/elements/addons/sc-top-sheet-views.js:
```diff
@@ -330,6 +338,31 @@ export class SCTopSheetViews extends LitLocalized(LitElement) {
+
+  // This componenet is _never_ disconnected the way the code is written today (2024-08-04)
+  // Still, adding this in case some day this component is properly unloaded.
+  disconnectedCallback() {
```
I was expecting this callback to be called when navigating from a sutta page to e.g. the homepage, but it wasn't called.  It seems that the top sheets are never unloaded?

ihongda replied:
Yes, this component is not actively uninstalled from the DOM.

khemarato commented on client/elements/menus/sc-action-items.js:
```diff
@@ -304,6 +304,13 @@ export class SCActionItems extends LitLocalized(LitElement) {
+
+  _handleKeydown(e) {
+    if (e.key === 'Escape') {
+      this.#closeAllTopSheetsOfTextPage();
```
While I'm here, I also added a keyboard shortcut to close all the top sheets: `esc`

khemarato commented on client/localization/elements/viewoption_en.json:
```diff
@@ -35,7 +35,7 @@
   "viewoption:sidenotes": "Show as sidenotes",
   "viewoption:textView": "Original text and translation",
-  "viewoption:textViewDescription": "Some texts on SuttaCentral are “segmented”; the original text and translation are approximately matched line by line. For such texts, see both root text and translation.",
+  "viewoption:textViewDescription": "Some texts on SuttaCentral are “segmented”; the original text and translation are approximately matched line by line. For such texts, see both root text and translation. Press <code class=\"key\">V</code> to toggle between the views.",
```
The other string changes were pulled in automatically by the build (from where?) except for this one which I changed to get the helptext updated.  What is the correct way to update this string, @sujato ?

thesunshade replied:
I think this has to be updated in `bilara-data` and then go through all its magic.
I'm wondering if this help text should get its own new segment.

khemarato replied:
@thesunshade - Can you take the lead on this upstream change? :pray:

thesunshade replied:
Sure. Just give me a poke when the shortcut is live if I don't notice.

khemarato reviewed (commented):
cc @thesunshade

thesunshade commented:
Thanks!!!!!

I am able to break it, though. 

1. Go to an alligned sutta, e.g. MN 5.
2. Cycle through the views by pressing `v`.
3. Click on the parent vagga in the breadcrumb trail
![image](https://github.com/user-attachments/assets/d58e472b-5087-45aa-87a4-d50e5a4c5ad0)
4. click on a new aligned translation, e.g. MN 6
5. Cycle through the views by pressing `v`.

**What doesn't happen:**
* the actual view does not change (meaning that the text itself does not change)

**What does happen**
* the snack bar announces the change
* the params in the url change

**What probably happens**
* the radio buttons change

HOWEVER… once you click on the "views" menu, then the change is applied and the new view is shown. After that, the `v` toggle works just fine.

Note, the `esc` key seems to work immediately, which probably makes sense. Opening the "info" panel doesn't get the view toggling working, which probably makes sense.

ihongda commented on client/elements/addons/sc-top-sheet-views.js:
```diff
@@ -330,6 +338,31 @@ export class SCTopSheetViews extends LitLocalized(LitElement) {
+    switch (event.key) {
+      case 'v':
+      case 'V':
+        var idx = this.textViewArray.findIndex(item => item.textView === this.selectedTextView);
```
Using `const` or `let` instead of `var` is more in line with the specification 🙏🏼

khemarato committed:
Actually working now

khemarato commented on client/elements/addons/sc-top-sheet-views.js:
```diff
@@ -332,6 +340,27 @@ export class SCTopSheetViews extends LitLocalized(LitElement) {
+        let idx = this.textViewArray.findIndex(item => item.textView === this.selectedTextView);
+        if (idx >= 0) {
+          idx = (idx + 1) % this.textViewArray.length;
+          this.#setTextBilaraPageIsTextOptionsMismatchSavedSettingsState();
```
@ihongda - this is the magic function call I needed to make it work.

ihongda replied:
👍🏼

ihongda reviewed (approved):
Thanks, Bhante 🙏🏼

thesunshade commented:
My suggestion would be to hold off on the interface changes for now. There are some folks who will start localizing things immediately and it's probably good to have further discussions on the best way to notify users of the hot-keys.

khemarato commented:
The string change here will get overwritten on the next `make` right?

thesunshade commented:
> The string change here will get overwritten on the next `make` right?

Yes, sorry. I was speaking more towards HongDa. I do believe that your change will be lost.

It's a discussion to be held probably in the issue, but I'm not even sure what Bhante had in mind. There will be a few that won't have an easy place to note in the existing interface.

BTW, I forgot to test what happens when you type a `v` when entering something in the search field. Did you?

khemarato committed:
remove string and fix search bug

khemarato commented:
Thanks @thesunshade - That was indeed getting captured. Somehow I thought it would automatically ignore `input` keypresses.  Anyway, that's now fixed and I've removed the string change. :)
